### PR TITLE
Storage buffer support part 1

### PR DESCRIPTION
**Pull request by Jhonnyg**
This PR adds support for creating storage buffers in the graphics adapter(s). No user-facing changes for now.

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/ShaderProgramBuilderTest.java [hunk 1]
```diff
@@ -145,6 +145,209 @@ private void doTest(boolean expectSpirv) throws Exception {
...
+        }
+
+        {
+            String fs_src =
+                "#version 430\n" +
+                "uniform sampler2D sampler_2d;\n" +
+                "uniform sampler2DArray sampler_2d_array;\n" +
+                "uniform samplerCube sampler_cube;\n" +
+                // TODO:
+                //"uniform samplerBuffer sampler_buffer;\n" +
```
Comment (Jhonnyg) [inline com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/ShaderProgramBuilderTest.java:307]: We don't need to support this right now, so it's just left as a note for the future

File: engine/graphics/src/graphics.cpp [hunk 2]
```diff
@@ -929,27 +934,114 @@ namespace dmGraphics
...
+            res.m_BindingFamily        = family;
+
+            if (res.m_Type.m_UseTypeIndex)
+                res.m_Type.m_TypeIndex = bindings[i].m_Type.m_Type.m_TypeIndex;
+            else
+                res.m_Type.m_ShaderType = bindings[i].m_Type.m_Type.m_ShaderType;
+        }
+    }
+
+    void CreateShaderMeta(ShaderDesc::Shader* ddf, ShaderMeta* meta)
```
Comment (Jhonnyg) [inline engine/graphics/src/graphics.cpp:960]: Imo we should migrate all the graphics adapters to use this format. I think we could simplify things in the renderer as well for getting uniform locations and attributes and whatnot. We could return the shader meta mapping instead, which contains everything the renderer needs to know.

File: engine/graphics/src/vulkan/graphics_vulkan.cpp [hunk 3]
```diff
@@ -4661,9 +4728,16 @@ namespace dmGraphics
...
                 {
-                    for (int i = 0; i < program_resource_binding.m_Res->m_BlockMembers.Size(); ++i)
+                    // TODO: Generic type lookup is not supported yet!
+                    // We can only support one level of indirection here right now
+                    assert(pgm_res.m_Res->m_Type.m_UseTypeIndex);
+                    const dmArray<ShaderResourceTypeInfo>& type_infos = *pgm_res.m_TypeInfos;
+                    const ShaderResourceTypeInfo& type_info = type_infos[pgm_res.m_Res->m_Type.m_TypeIndex];
+
+                    const uint32_t num_members = type_info.m_Members.Size();
+                    for (int i = 0; i < num_members; ++i)
```
Comment (Jhonnyg) [inline engine/graphics/src/vulkan/graphics_vulkan.cpp:4740]: This is the same as we had before, but now with the type tables we can support any type of generic structural data in shaders!

**Review by JCash — DISMISSED** [review]
A lot of stuff, but lgtm!

**Pull request merged**